tests/unit/lms/services/grouping_test.py

```diff
@@ -15,100 +14,80 @@ class TestGroupingService:
     CONTEXT_ID = "context_id"
     TOOL_CONSUMER_INSTANCE_GUID = "t_c_i_guid"
 
-    def test_upsert_inserts(self, svc, db_session):
-        # Start with no groupings
-        assert not db_session.query(Grouping).count()
+    def test_upsert_with_parents_inserts(self, svc, db_session):
+        course = factories.Course()
+        application_instance = factories.ApplicationInstance()
 
-        test_grouping = Grouping(
-            application_instance=factories.ApplicationInstance(),
-            authority_provided_id="ID",
+        # Start with no CanvasGroup
+        assert not db_session.query(CanvasGroup).count()
+
+        test_grouping = svc.upsert_with_parent(
+            tool_consumer_instance_guid=application_instance.tool_consumer_instance_guid,
             lms_id="lms_id",
             lms_name="lms_name",
+            parent=course,
+            type_=Grouping.Type.CANVAS_GROUP,
         )
 
-        svc.upsert(test_grouping)
-
-        assert db_session.query(Grouping).one() == test_grouping
+        assert db_session.query(CanvasGroup).one() == test_grouping
 
-    def test_upsert_updates(self, svc, db_session):
-        grouping = factories.Grouping()
+    def test_upsert_with_parents_updates(self, svc, db_session, application_instance):
```

Comment:
This test doesn't seem right. If I break the code so that it actually doesn't update anything the test still passes:

```diff
diff --git a/lms/services/grouping.py b/lms/services/grouping.py
index 48d26486..fc647c03 100644
--- a/lms/services/grouping.py
+++ b/lms/services/grouping.py
@@ -73,10 +73,6 @@ class GroupingService:
         )
         if not db_grouping:
             self._db.add(grouping)
-        else:
-            # Update any fields that might have changed
-            db_grouping.lms_name = grouping.lms_name
-            db_grouping.extra = grouping.extra
 
         return db_grouping or grouping
 
```

It looks like the test code itself actually updates the grouping:

https://github.com/hypothesis/lms/blob/fa8ee0f917db775b9989c05d6d0c8c652067d751/tests/unit/lms/services/grouping_test.py#L49-L50

Here's a replacement test that works. It works by calling `upsert_with_parent()` twice which in a way isn't ideal, but it leads to a simple test that does fail if the code is wrong:

```python
    def test_upsert_with_parents_updates(self, svc, db_session, application_instance):
        course = factories.Course()
        tool_consumer_instance_guid = course.application_instance.tool_consumer_instance_guid
        lms_id = "lms_id"
        old_name = "old_name"
        old_extra = {"extra": "old"}
        new_name = "new_name"
        new_extra = {"extra": "new"}

        # Insert a new grouping.
        svc.upsert_with_parent(
            tool_consumer_instance_guid=tool_consumer_instance_guid,
            lms_id=lms_id,
            lms_name=old_name,
            parent=course,
            type_=Grouping.Type.CANVAS_GROUP,
            extra=old_extra,
        )

        # Update the previously-inserted grouping with new values.
        grouping = svc.upsert_with_parent(
            tool_consumer_instance_guid=tool_consumer_instance_guid,
            lms_id=lms_id,
            lms_name=new_name,
            parent=course,
            type_=Grouping.Type.CANVAS_GROUP,
            extra=new_extra,
        )

        # upsert_with_parent() should return a Grouping that has the updated values.
        assert grouping.lms_name == new_name
        assert grouping.extra == new_extra

        # The Grouping's values should have been updated in the DB as well.
        db_grouping = db_session.query(CanvasGroup).one()
        assert db_grouping.lms_name == new_name
        assert db_grouping.extra == new_extra
```

Reply:
Great thanks for the new test version.

The one I had broke (as in it wasn't testing the real  code) after removing `make_transient(grouping)`